File: examples/Ruby/syncsub.rb
#
# Synchronized subscriber
#

require 'rubygems'
require 'ffi-rzmq'

context = ZMQ::Context.new

# First, connect our subscriber socket
subscriber = context.socket(ZMQ::SUB)
subscriber.connect("tcp://localhost:5561")
subscriber.setsockopt(ZMQ::SUBSCRIBE,"")

# 0MQ is so fast, we need to wait a while...
sleep(1)

# Second, synchronize with publisher
synclient = context.socket(ZMQ::REQ)
synclient.connect("tcp://localhost:5562")

# - send a synchronization request
synclient.send_string("")

# - wait for synchronization reply
synclient.recv_string('')

# Third, get our updates and report how many we got
update_nbr=0
loop do 
  subscriber.recv_string(string = '')
  break if string == "END"
  update_nbr+=1
end

puts "Received #{update_nbr} updates"

File: examples/Haskell/syncsub.hs
{-# LANGUAGE OverloadedStrings #-}

--  Synchronized subscriber

module Main where

import Control.Concurrent
import Data.Function
import System.ZMQ4.Monadic
import Text.Printf

main :: IO ()
main = runZMQ $ do
    -- First, connect our subscriber socket
    subscriber <- socket Sub
    connect subscriber "tcp://localhost:5561"
    subscribe subscriber ""

    -- 0MQ is so fast, we need to wait a while...
    liftIO $ threadDelay 1000000

    -- Second, synchronize with the publisher
    syncclient <- socket Req
    connect syncclient "tcp://localhost:5562"

    -- Send a synchronization request
    send syncclient [] ""

    -- Wait for a synchronization reply
    receive syncclient

    let -- go :: (Int -> ZMQ z Int) -> Int -> ZMQ z Int
        go loop = \n -> do
            string <- receive subscriber
            if string == "END"
                then return n
                else loop (n+1)

    -- Third, get our updates and report how many we got
    update_nbr <- fix go (0 :: Int)
    liftIO $ printf "Received %d updates\n" update_nbr
